.github/workflows/build.yml

```diff
@@ -0,0 +1,471 @@
+          mkdir -p out
+      - name: Build Linux distribution packages
+        uses: hashicorp/package@v1
+        with:
+          name: "terraform"
+          description: "Terraform enables you to safely and predictably create, change, and improve infrastructure. It is an open source tool that codifies APIs into declarative configuration files that can be shared amongst team members, treated as code, edited, reviewed, and versioned."
+          arch: ${{ matrix.arch }}
+          version: ${{ env.version }}
+          maintainer: "HashiCorp"
+          homepage: "https://terraform.io/"
+          license: "MPL-2.0"
+          binary: "dist/terraform"
+          deb_depends: "git"
+          rpm_depends: "git"
+      - name: Gather Linux distribution package filenames
+        run: |
+          echo "RPM_PACKAGE=$(basename out/*.rpm)" >> $GITHUB_ENV
+          echo "DEB_PACKAGE=$(basename out/*.deb)" >> $GITHUB_ENV
+      - name: "Save .rpm package"
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ env.RPM_PACKAGE }}
+          path: out/${{ env.RPM_PACKAGE }}
+      - name: "Save .deb package"
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ env.DEB_PACKAGE }}
+          path: out/${{ env.DEB_PACKAGE }}
+
+  # TODO: homebrew packages for macOS
+  #package-homebrew:
+  #  name: Build Homebrew package for darwin_${{ matrix.arch }}
+  #  runs-on: macos-latest
+  #  needs:
+  #    - get-product-version
+  #    - build
+  #  strategy:
+  #    matrix:
+  #      arch: ["amd64", "arm64"]
+  #    fail-fast: false
+  # ...
+
+  package-docker:
+    name: Build Docker image for linux_${{ matrix.arch }}
+    runs-on: ubuntu-latest
+    needs:
+      - get-product-version
+      - build
+    strategy:
+      matrix:
+        arch: ["amd64"]
+      fail-fast: false
+
+    env:
+      repo: ${{github.event.repository.name}}
+      version: ${{needs.get-product-version.outputs.product-version}}
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Build Docker images
+        uses: hashicorp/actions-docker-build@v1
+        with:
+          version: ${{env.version}}
+          target: default
+          arch: ${{matrix.arch}}
+          dockerfile: .github/workflows/build-Dockerfile
+          tags: |
+            docker.io/hashicorp/${{env.repo}}:${{env.version}}
+            986891699432.dkr.ecr.us-east-1.amazonaws.com/hashicorp/${{env.repo}}:${{env.version}}
+
+  e2etest-build:
+    name: Build e2etest for ${{ matrix.goos }}_${{ matrix.goarch }}
+    runs-on: ubuntu-latest
+    needs: ["get-go-version"]
+    strategy:
+      matrix:
+        # We build test harnesses only for the v1.0 Compatibility Promises
+        # supported platforms. Even within that set, we can only run on
+        # architectures for which we have GitHub Actions runners available,
+        # which is currently only amd64 (x64).
+        # TODO: GitHub Actions does support _self-hosted_ arm and arm64
```

Comment:
There's also qemu. It is significantly slower as it's cpu instruction emulation. But for low-cpu tasks it works great.

Just use this github action and then you can exec both amd64 and arm64 binaries https://github.com/docker/setup-qemu-action

Reply:
I can have a go and updating package-docker to build arm64 docker image and and e2etest-* to test arm64 binary using amd64 test harness if that sounds ok?

Reply:
Hi @morganchristiansson,

Thanks for the tip!

I believe the packaging helper action we're using here already has some support for packaging across architectures -- I think I saw it work when I was first experimenting. However, I intentionally reduced it to just what our current release pipeline is capable of so that we wouldn't get potentially blocked by irrelevant failures in the meantime, given that we're currently using this only to get early feedback on build/e2etest problems.

I'd like to wait to add new architectures here until our release engineering team is ready to work with us on feeding this into the rest of the new pipeline, so that we can make sure whatever we do here is consistent with how other projects using the same pipeline will do it.

Reply:
Btw I had a look at https://github.com/hashicorp/actions-docker-build it already uses `docker build buildx --platform $PLATFORM` but $PLATFORM is only used internally not settable from action parameters. https://github.com/hashicorp/actions-docker-build/blob/v1/action.yml#L222-L224`

If arm64 binary is being built here? It will be easy to add the binary to arm64 docker image like this and will run well...